runtime/major_gc.c

```diff
@@ -359,91 +524,76 @@ static value* mark_ephe_aux (value *gray_vals_ptr, intnat *work,
-  start = current_index;
-  while (work > 0){
-    if (v == 0 && gray_vals_ptr > gray_vals){
-      CAMLassert (start == 0);
-      v = *--gray_vals_ptr;
-      CAMLassert (Is_black_val (v));
-#ifdef NO_NAKED_POINTERS
-      if (Tag_val(v) == Closure_tag) {
-        /* Skip the code pointers and integers at beginning of closure;
-           start scanning at the first word of the environment part. */
-        start = Start_env_closinfo(Closinfo_val(v));
-        CAMLassert(start <= Wosize_val(v));
+
+  while (1){
+    int can_mark = 0;
+
+    if (me.offset == me_end) {
+      if (stk->count > 0)
+      {
+        me = stk->stack[--stk->count];
+        me_end = Wosize_val(me.block);
+        can_mark = 1;
       }
-#endif
     }
-    if (v != 0){
-      hd = Hd_val(v);
-      CAMLassert (Is_black_hd (hd));
-      size = Wosize_hd (hd);
-      end = start + work;
-      if (Tag_hd (hd) < No_scan_tag){
-        start = size < start ? size : start;
-        end = size < end ? size : end;
-        CAMLassert (end >= start);
-        CAML_EVENTLOG_DO({
-          slice_fields += end - start;
-          if (size > end)
-            CAML_EV_COUNTER (EV_C_MAJOR_MARK_SLICE_REMAIN, size - end);
-        });
-        for (i = start; i < end; i++){
-          gray_vals_ptr = mark_slice_darken(gray_vals_ptr,v,i,
-                                            /*in_ephemeron=*/ 0,
-                                            &slice_pointers);
-        }
-        if (end < size){
-          work = 0;
-          start = end;
-          /* [v] doesn't change. */
-          CAMLassert (Is_black_val (v));
-        }else{
-          CAMLassert (end == size);
-          Hd_val (v) = Blackhd_hd (hd);
-          work -= Whsize_wosize(end - start);
-          start = 0;
-          v = 0;
-        }
-      }else{
-        /* The block doesn't contain any pointers. */
-        CAMLassert (start == 0);
-        Hd_val (v) = Blackhd_hd (hd);
-        work -= Whsize_wosize(size);
-        v = 0;
+    else
+    {
+      can_mark = 1;
+    }
+
+    if (work <= 0) {
+      if( can_mark ) {
+        mark_stack_push(stk, me, NULL);
+      }
+      break;
+    }
+
+    if( can_mark ) {
+      CAMLassert(Is_block(me.block) &&
+                Is_black_val (me.block) &&
+                Tag_val(me.block) < No_scan_tag);
+
+      mark_slice_darken(stk, me.block, me.offset++, /*in_ephemeron=*/ 0,
+                                              &slice_pointers, &work);
+
+      work--;
+
+      if( me.offset == me_end ) {
+        work--; /* Include header word */
+      }
+    } else if( redarken_first_chunk != NULL ) {
+      /* There are chunks that need to be redarkened because we
+         overflowed our mark stack */
+      if( redarken_chunk(redarken_first_chunk, stk) ) {
+        do {
+          redarken_first_chunk = Chunk_next(redarken_first_chunk);
+        } while( redarken_first_chunk != NULL
+                  && !Chunk_requires_redarken(redarken_first_chunk) );
```

Comment:
```suggestion
        redarken_first_chunk = Chunk_next(redarken_first_chunk);
```

I think you can weaken the invariant of `redarken_first_chunk` from "this is the first chunk which requires redarkening" to "all the chunks that need redarkening appear after `redarken_first_chunk`". Then, you no longer need this loop, and the `Chunk_requires_redarken` field becomes useless.

The `while(1)` loop will naturally do the work of finding the next chunk that need redarkening, and the `Chunk_requires_redarken` check will be natural in `redarken_chunk` thanks to the sentinels you use.

Reply:
This works and is a good idea. It is implemented in one of the most recent commits.